[Cases] Toggle columns in the cases table

adcoelho opened the pull request:
Connected to #167617

This PR is being merged into a feature branch

## Summary

This PR allows users to configure which columns are displayed in the cases table and in which order.

The configuration is persisted in local storage and is application-specific.

Custom fields can also be selected.

https://github.com/elastic/kibana/assets/1533137/da386a6c-b2d9-4a86-948f-79903163ff3c

## Coming later in the Feature Branch

The scope of this PR is limited to the Summary above. There are still some things under consideration and some coming later. These are:
- The display of the toggle in the table. Right now it looks clickable when it isn't. I will discuss it with @mdefazio and pick a different design.
- Scrolling inside the popover
- Search inside the popover
- `Show all` and `Hide all` functionality
- In the final PR I will play with the column widths and try to find the best size for each attribute.

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

elasticmachine commented:
Pinging @elastic/response-ops-cases (Feature:Cases)

adcoelho commented on x-pack/plugins/cases/public/components/all_cases/columns_popover.test.tsx:
```diff
@@ -0,0 +1,71 @@
+    });
+  });
+
+  it.skip('drag and rop calls onSelectedColumnsChange with the right params', async () => {});
```
Any ideas? I've only seen this being done with _enzyme_ to `simulate` the event. I tried using https://testing-library.com/docs/dom-testing-library/api-events/#createeventeventname but it did not work. I looked into the `eui` documentation but couldn't find any relevant drag-and-drop testing.

jcger commented:
I've had a chance to review the recent changes, and while they look impressive, I did come across a few issues during the QA process:

1. What about making it sticky instead of letting it disappear on scroll?
<details>
<summary>Video</summary>
<video src="https://github.com/elastic/kibana/assets/17549662/7da62d85-8a0b-4cf4-8a34-57dbd09d3b72"style="max-height:640px; min-height: 200px">
  </video>
</details>

2. The margin between the last two isn't ideal. Maybe we need an ellipsis there?
<details>
<summary>Image</summary>
<img src="https://github.com/elastic/kibana/assets/17549662/ec650812-0287-4e68-ab6e-7514c20c0564" width=350/>
</details>

3. Fields appearing/disappear, not sure what the right behaviour would be
<details>
<summary>Video</summary>
<video src="https://github.com/elastic/kibana/assets/17549662/f60c66cb-41c4-484e-a298-c62e370750bd"style="max-height:640px; min-height: 200px">
</details>

4. It looks like there is a render before it loads the columns from the local storage causing a layout shift. Would be nice if we could get rid off it
<details>
<summary>Video</summary>
<video src="https://github.com/elastic/kibana/assets/17549662/c6f4e6cd-3464-4e61-8e9f-33525e5ddabb"style="max-height:640px; min-height: 200px">
</details>

adcoelho commented:
@jcger I'll respond to them 1 by 1 in separate comments 😅

> What about making it sticky instead of letting it disappear on scroll?

I am afraid this is the `eui` default behavior. If the anchor(in this case the columns button) the `popover` is tied to is scrolled out of the page, the `popover` also goes away.

[All popovers behave like that in the documentation.](https://eui.elastic.co/#/layout/popover)

One alternative would be [the popover on a fixed element](https://eui.elastic.co/#/layout/popover#popover-on-a-fixed-element) but then the columns button would "hover" over the table and I don't think it looks that much better to be honest.

<details><summary>See here.</summary>
<video src="https://github.com/elastic/kibana/assets/1533137/317e7530-b5dc-4a17-8b0b-7102af075986" style="max-height:640px; min-height: 200px"></video>
</details>


_If the column headers are not on the page, I think it is fine the popover isn't either._

adcoelho commented:
@jcger follow up

> 2. The margin between the last two isn't ideal. Maybe we need an ellipsis there?

Fixed, thank you for noticing it 🙌  (will push soon)

> 4. It looks like there is a render before it loads the columns from the local storage causing a layout shift. Would be nice if we could get rid of it.

This was a very nice catch. I started by looking into the local storage but turns out the reason is actually `useGetCaseConfiguration`.

What is happening is that all non-custom field columns render immediately and when `useGetCaseConfiguration` returns the custom field columns are added. I am fixing it now 😉

jcger commented:
> @jcger I'll respond to them 1 by 1 in separate comments 😅
> 
> > What about making it sticky instead of letting it disappear on scroll?
> 
> I am afraid this is the `eui` default behavior. If the anchor(in this case the columns button) the `popover` is tied to is scrolled out of the page, the `popover` also goes away.
> 
> [All popovers behave like that in the documentation.](https://eui.elastic.co/#/layout/popover)
> 
> One alternative would be [the popover on a fixed element](https://eui.elastic.co/#/layout/popover#popover-on-a-fixed-element) but then the columns button would "hover" over the table and I don't think it looks that much better to be honest.
> 
> See here.
> _If the column headers are not on the page, I think it is fine the popover isn't either._

I thought you might be able to use `position: sticky` https://developer.mozilla.org/en-US/docs/Web/CSS/position

adcoelho commented:
@jcger 

I finally fixed

> 4. It looks like there is a render before it loads the columns from the local storage causing a layout shift. It would be nice if we could get rid of it.

In the end, it wasn't about the local storage or a delay in loading the permissions as I thought initially.

The list in local storage is usually merged with the configuration in case some custom field was created or any new field was added.

What I forgot was that those fields should not be included if the user doesn't have permission to see them(`!canDisplay`) 😅 

I fixed that in the latest commit and added a test for the scenario.

jcger commented on x-pack/plugins/cases/public/components/all_cases/utils.ts:
```diff
@@ -28,3 +34,52 @@ export const parseUrlQueryParams = (parsedUrlParams: ParsedUrlQueryParams): Part
+  selectedColumns: CasesColumnSelection[];
+  casesColumnsConfig: CasesColumnsConfiguration;
+}): CasesColumnSelection[] => {
+  // selectedColumns is the master
```
you could get rid of this comment, the code is self explanatory enough to make this unnecessary

jcger commented on x-pack/plugins/cases/public/components/custom_fields/text/get_column.tsx:
```diff
@@ -0,0 +1,39 @@
+import type { CaseUI } from '../../../containers/types';
+import { getEmptyTagValue } from '../../empty_value';
+
+export const getColumn = ({
```
maybe we could rename this to know from just seeing the file name what is it, get_column might be to generic. What about something like get_eui_table_column_props or something similar?

adcoelho commented on x-pack/plugins/cases/public/components/all_cases/use_cases_columns_configuration.tsx:
```diff
@@ -0,0 +1,120 @@
+      canDisplay: canDisplayDefault,
+    },
+    assignCaseAction: {
+      field: '',
```
Tbh I plan to make the `field` optional in the `CasesColumnsConfiguration` in a future PR.

Not all `eui` table column types require the field property. If you have a "computed column" it won't match a single attribute and the whole Case is passed to the `render`(column) callback. `assignCaseAction` is one of those so the field is not needed here.

However, further down the line in the column popover, we still assume a field exists, but I will find a workaround.

cnasikas commented on x-pack/plugins/cases/public/components/all_cases/use_cases_columns_selection.test.tsx:
```diff
@@ -0,0 +1,140 @@
+import { DEFAULT_CASES_TABLE_COLUMNS } from '../../../common/constants';
+
+jest.mock('./use_cases_columns_configuration');
+jest.mock('./utils');
```
I don't think we should mock the `utils`. There is no API calls or similar to justify it.

cnasikas commented on x-pack/plugins/cases/public/components/custom_fields/text/get_eui_table_column.tsx:
```diff
@@ -0,0 +1,39 @@
+}): EuiTableComputedColumnType<CaseUI> => ({
+  name: label,
+  render: (theCase: CaseUI) => {
+    const index = theCase.customFields.findIndex(
```
nit: You can use `find`

cnasikas commented on x-pack/plugins/cases/public/components/custom_fields/text/get_eui_table_column.test.tsx:
```diff
@@ -0,0 +1,65 @@
+
+    theCase.customFields = [{ key, type: CustomFieldTypes.TEXT, value }];
+
+    render(<TestProviders>{column.render(theCase)}</TestProviders>);
```
nit: Better to use the `appMockRenderer`

cnasikas commented on x-pack/plugins/cases/public/components/custom_fields/toggle/get_eui_table_column.tsx:
```diff
@@ -0,0 +1,62 @@
+
+    if (index === -1) {
+      return (
+        <EuiSwitch
```
What is the latest status from @mdefazio about how we show this?

adcoelho replied:
I followed the comment [here](https://github.com/elastic/kibana/issues/167617#issuecomment-1781081636) and let icons. I like how it looks.

cnasikas replied:
Nice!

cnasikas commented on x-pack/plugins/cases/public/components/custom_fields/toggle/get_eui_table_column.tsx:
```diff
@@ -0,0 +1,62 @@
+      );
+    }
+
+    if (theCase.customFields[index].type !== CustomFieldTypes.TOGGLE) {
```
I think we can safely omit this check here.

cnasikas commented on x-pack/plugins/cases/public/components/custom_fields/toggle/get_eui_table_column.tsx:
```diff
@@ -0,0 +1,62 @@
+  label: string;
+}): EuiTableComputedColumnType<CaseUI> => ({
+  name: label,
+  render: (theCase: CaseUI) => {
```
I think the whole render function can be simplified as:

```

const customField = theCase.customFields.find((element) => element.key === key && element.value != null); <--- we need the null check here. It can be set through the API.

const isChecked = Boolean(customField?.value)

return (
      <EuiSwitch
        data-test-subj={`toggle-custom-field-column-view-${key}`}
        label={''}
        checked={isChecked}
        onChange={() => {}}
        compressed
      />
    );
```

cnasikas commented on x-pack/plugins/cases/public/components/custom_fields/types.ts:
```diff
@@ -30,6 +36,7 @@ export interface CustomFieldType<T extends CaseUICustomField> {
 export type CustomFieldFactory<T extends CaseUICustomField> = () => {
   id: string;
   label: string;
+  getEuiTableColumn: (params: { key: string; label: string }) => EuiTableComputedColumnType<CaseUI>;
```
I am a bit sceptical with this. Custom fields now have knowledge to the structure of cases which they shouldn't. What about we pass only the custom field to the render method? Example: `getEuiTableColumn: (params: { key: string; label: string }) => EuiTableComputedColumnType<CaseUICustomField>;` This way custom fields do not have to search on the case and are not aware of the structure. We can talk offline if you want of how to do it.

adcoelho replied:
I can pass `customField` as the field in the column and the render function will get the `customFields` array as a parameter, but I don't think we can get out of iterating over that array. Is this what you mean?

cnasikas replied:
I think we can. Let's talk about it offline.

adcoelho replied:
Discussed offline, found a solution.

cnasikas commented on x-pack/test/functional_with_es_ssl/apps/cases/group2/list_view.ts:
```diff
@@ -658,5 +658,43 @@ export default ({ getPageObject, getService }: FtrProviderContext) => {
+      after(async () => {
+        await cases.api.deleteAllCases();
+        await cases.casesTable.waitForCasesToBeDeleted();
+        await browser.clearLocalStorage();
```
nice!

cnasikas commented on x-pack/plugins/cases/public/components/custom_fields/text/get_eui_table_column.tsx:
```diff
@@ -0,0 +1,39 @@
+import type { CaseUI } from '../../../containers/types';
+import { getEmptyTagValue } from '../../empty_value';
+
+export const getEuiTableColumn = ({
```
Should we set a `width`?

adcoelho replied:
I set one for now but will still play more with this in the last PR to go over the overall look of the table.

cnasikas commented on x-pack/plugins/cases/public/components/custom_fields/text/get_eui_table_column.tsx:
```diff
@@ -0,0 +1,39 @@
+
+    if (index !== -1) {
+      return (
+        <p data-test-subj={`text-custom-field-column-view-${key}`}>
```
Should we truncate the text?

**With truncation**

<img width="2282" alt="Screenshot 2023-10-30 at 12 54 29 PM" src="https://github.com/elastic/kibana/assets/7871006/99022ac2-465b-4442-9d36-f6e2d2efa421">

**Without truncation**

<img width="2274" alt="Screenshot 2023-10-30 at 12 54 04 PM" src="https://github.com/elastic/kibana/assets/7871006/82e98fd3-c0c8-49cc-94d8-79fc429316ed">

cnasikas commented on x-pack/plugins/cases/public/components/custom_fields/toggle/get_eui_table_column.tsx:
```diff
@@ -0,0 +1,62 @@
+import { CustomFieldTypes } from '../../../../common/types/domain';
+import { getEmptyTagValue } from '../../empty_value';
+
+export const getEuiTableColumn = ({
```
Should we set a `width`? It is a toogle and the size will always be small.

mdefazio replied:
An aside regarding column widths and the potential craziness of the table as more columns are added...

I think we can work towards having the table scroll horizontally.  Otherwise, we will constantly be battling this.  
So perhaps setting columns to auto (with a max-width) and just let the table overflow.  There weren't a11y concerns with this approach (we would just need to make sure we provide appropriate keyboard access).  But I think finding magic widths here is going to be a losing battle.  

Should this be built into the EUI table? Perhaps.  Seems like the team there is +1 for this, but this obviously is beyond this PR's scope.  

Example gif (with an option to pin a column or columns).  
![Screenshot--2023-10-26--Table Designs - UI Updates](https://github.com/elastic/kibana/assets/3756330/25a06ce0-9664-46d4-a7cf-083cac0a2408)

cnasikas replied:
Thanks, @mdefazio for the input. I totally agree that is impossible to find a good balance with the widths. Horizontally scrolling seems the way to go. @adcoelho What do you think about contributing to the EUI table as a stretch goal after we are done with the feature?

cnasikas reviewed (commented):
Great work! I really liked the feature and how it looks 🚀. I found the PR a bit big with a lot of new functionality to review. I left some comments.

adcoelho commented on x-pack/plugins/cases/public/components/all_cases/table.tsx:
```diff
@@ -57,6 +65,9 @@ export const CasesTable: FunctionComponent<CasesTableProps> = ({
   deselectCases,
+  selectedColumns,
+  onSelectedColumnsChange,
+  isFetchingColumns,
```
I had that initially and thought it was confusing to rename `isFetching` to `isLoading` 😄

Now I am not sure anymore.

adcoelho commented on x-pack/plugins/cases/public/components/all_cases/use_cases_columns.test.tsx:
```diff
@@ -129,77 +153,22 @@ describe('useCasesColumns ', () => {
   });
 
-  it('returns the assignees column without the width specified when in the modal view', async () => {
+  it('only returns selected columns', async () => {
```
In the previous test all input columns have `isChecked: true`. Here some have `isChecked: false`.

adcoelho commented on x-pack/plugins/cases/public/components/all_cases/use_cases_columns.tsx:
```diff
@@ -322,57 +306,76 @@ export const useCasesColumns = ({
+    });
+  } else {
+    selectedColumns.forEach(({ field, isChecked }) => {
+      if (field in columnsDict && isChecked) {
```
That check is done when retrieving `selectedColumns`. If they cannot be displayed they will never circulate around the components in the `selectedColumns` array.

cnasikas replied:
What if this assumption changes in the future or the hook is used in a different context? IMO the `casesColumnsConfig` is the source of truth and we should rely on it without assumptions outside of the hook.

adcoelho replied:
> IMO the casesColumnsConfig is the source of truth and we should rely on it without assumptions outside of the hook.

Exactly, and it was used to build `selectedColumns`. The hook already uses the configuration to exclude `canDisplay: false` so in practice we would be doing the same thing twice.

adcoelho replied:
Nah, I can include it. `useCasesColumns` already has the configuration in it so it is just an extra check in this `if` 👍

adcoelho commented on x-pack/plugins/cases/public/components/all_cases/columns_popover.tsx:
```diff
@@ -0,0 +1,143 @@
+                          onChange={(e) => toggleColumn({ field, isChecked: e.target.checked })}
+                          compressed
+                          labelProps={{
+                            style: {
```
I tried `className: 'eui-textTruncate'` and `css: css'${euiTextTruncate()}'` neither worked. Maybe because this is part of the `labelProps`.

cnasikas replied:
So `labelProps: { className: 'eui-textTruncate' }` does not work correctly, right?

adcoelho replied:
yeah, it doesnt truncate :/

cnasikas replied:
😞

js-jankisalvi reviewed (commented):
Verified locally, works as expected 👍 
I especially like how smooth column selection and deselection reflects on table!! 😍 
Great job! 🎉 

I had small concern about popover in small device. Not sure we need to take care of popover in mobile devices, but it looks like this: 
![image](https://github.com/elastic/kibana/assets/117571355/33a68a72-fd47-4aa3-b133-96d702a99e85)

adcoelho commented on x-pack/plugins/cases/common/constants/index.ts:
```diff
@@ -164,6 +164,36 @@ export const DELETE_CASES_CAPABILITY = 'delete_cases' as const;
+
+// We just need the default ids(field) and isChecked.
+// The name comes from the cases columns configuration.
+export const DEFAULT_CASES_TABLE_COLUMNS: CasesColumnSelection[] = [
```
I wanted the default selection to be `CasesColumnSelection` but in practice we really just need the `key/field/id` and the `isChecked` because this gets merged with the configuration in `x-pack/plugins/cases/public/components/all_cases/use_cases_columns_selection.tsx`.

cnasikas replied:
Maybe this configuration can be the source of truth about the naming as it is static and cannot change dynamically.

adcoelho replied:
But it is not complete, does not include fields like alerts and comments e.g.
